setuptools: Support multiple builders

stephenfin opened the pull request:
Subject: Support multiple builders in setuptools integration

### Feature or Bugfix
- Feature

### Purpose

At present, the 'builder' option for the setuptools integration only
supports a single output format, typically HTML, like so:

    [build_sphinx]
    builder = man

There is value in being able to specify multiple format, like so:

    [build_sphinx]
    builder = html man

Make this possible.

### Detail
- N/A

### Relates
- N/A

stephenfin committed:
setuptools: Support multiple builders

At present, the 'builder' option for the setuptools integration only
supports a single output format, typically HTML, like so:

  [build_sphinx]
  builder = man

There is value in being able to specify multiple format, like so:

  [build_sphinx]
  builder = html man

Make this possible.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on sphinx/setup_command.py:
```diff
@@ -144,15 +145,19 @@ def finalize_options(self):
-        self.mkpath(self.builder_target_dir)  # type: ignore
+        self.builder_target_dirs = [
+            (builder, os.path.join(self.build_dir, builder))
+            for builder in self.builder]  # type: List[(str, str)]
```
This type annotation is invalid. You should use `List[Tuple[str, str]]` instead.
https://travis-ci.org/sphinx-doc/sphinx/jobs/205824862#L594

BTW, I think the second element of the tuple is a unicode string.
Certainly `builder` is raw string, but `self.build_dir` is a unicode (because we can build docs on path including multibytes). As a result, the joined string becomes a unicode string.

stephenfin replied:
Done

tk0miya commented on sphinx/setup_command.py:
```diff
@@ -144,15 +145,19 @@ def finalize_options(self):
             self.config_dir = self.source_dir
         self.config_dir = abspath(self.config_dir)
 
+        self.ensure_string_list('builder')
```
I'm sorry, but could you add `# type: ignore` label?
It seems mypy does not know `ensure_string_list()` unfortunately. 
https://travis-ci.org/sphinx-doc/sphinx/jobs/205824862#L593

stephenfin replied:
Done

tk0miya reviewed (changes requested):
LGTM with nits.

tk0miya commented on sphinx/setup_command.py:
```diff
@@ -144,15 +145,19 @@ def finalize_options(self):
-        self.mkpath(self.builder_target_dir)  # type: ignore
+        self.builder_target_dirs = [
+            (builder, os.path.join(self.build_dir, builder))
+            for builder in self.builder]  # type: List[Tuple[str, str]]
```
Oh, I'm very sorry. I'd forgotten to say you that you need to import `Tuple` like following:
```
 if False:
     # For type annotation
+    from typing import Any, Tuple  # NOQA
-    from typing import Any  # NOQA
```
https://travis-ci.org/sphinx-doc/sphinx/jobs/206202604#L591

stephenfin replied:
Weird - I'm running this locally (`tox -e py36`) and it's passing. Sorry about that

tk0miya replied:
Now type verification is invoked outside of tox: `make type-check`. style check (flake8) is also. Sorry for confusing

stephenfin committed:
tox: Add 'mypy' target

Like the recently added 'docs' target, this is easier than creating a
virtualenv, installing deps and running make.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented:
The test was broken, but it was caused by troubles in AWS.
Now I re-run tests manually, and it got succeeded.

tk0miya commented:
Merged. Thank you for contribution!
